File: bad_code.py
import os
import pickle
import random
import hashlib
import sqlite3
from flask import Flask, request
import subprocess

# Security Vulnerabilities

def sql_injection_vulnerable(user_input):
    """SQL Injection vulnerability"""
    conn = sqlite3.connect('database.db')
    cursor = conn.cursor()
    # SQL Injection - never do this
    query = f"SELECT * FROM users WHERE username = '{user_input}'"
    cursor.execute(query)
    return cursor.fetchall()

def command_injection(user_input):
    """Command injection vulnerability"""
    # Command injection risk
    os.system("ls " + user_input)
    subprocess.call("ping " + user_input, shell=True)

def hardcoded_credentials():
    """Hardcoded credentials - security issue"""
    password = "admin123"  # Hardcoded password
    api_key = "sk-1234567890abcdef"  # Hardcoded API key
    secret_token = "my_secret_token_12345"
    return password, api_key

def weak_cryptography(data):
    """Weak cryptographic algorithm"""
    # MD5 is weak and should not be used
    return hashlib.md5(data.encode()).hexdigest()

def insecure_random():
    """Using insecure random number generator"""
    # random module is not cryptographically secure
    return random.randint(1, 1000000)

def pickle_deserialization(data):
    """Insecure deserialization"""
    # Pickle is unsafe with untrusted data
    return pickle.loads(data)

def path_traversal(filename):
    """Path traversal vulnerability"""
    # No validation - can access any file
    with open("/data/" + filename, 'r') as f:
        return f.read()

app = Flask(__name__)

@app.route('/eval')
def dangerous_eval():
    """Code injection via eval"""
    user_code = request.args.get('code')
    return str(eval(user_code))  # Never use eval with user input

@app.route('/exec')
def dangerous_exec():
    """Code injection via exec"""
    user_code = request.args.get('code')
    exec(user_code)  # Never use exec with user input
    return "Executed"

# Code Smells

def too_many_parameters(a, b, c, d, e, f, g, h, i, j):
    """Too many parameters"""
    return a + b + c + d + e + f + g + h + i + j

def cognitive_complexity(x, y, z):
    """High cognitive complexity"""
    if x > 0:
        if y > 0:
            if z > 0:
                if x > y:
                    if y > z:
                        if x > z:
                            if x + y > z:
                                if x * y > z:
                                    return True
    return False

def duplicated_code_1():
    """Duplicated code block 1"""
    name = "John"
    age = 30
    email = "john@example.com"
    print(f"Name: {name}")
    print(f"Age: {age}")
    print(f"Email: {email}")
    print("Processing complete")

def duplicated_code_2():
    """Duplicated code block 2"""
    name = "Jane"
    age = 25
    email = "jane@example.com"
    print(f"Name: {name}")
    print(f"Age: {age}")
    print(f"Email: {email}")
    print("Processing complete")

def magic_numbers(price):
    """Magic numbers - should use constants"""
    tax = price * 0.18  # Magic number
    discount = price * 0.15  # Magic number
    shipping = price * 0.05  # Magic number
    return price + tax - discount + shipping

def unused_variables():
    """Unused variables"""
    unused_var = 42
    another_unused = "hello"
    result = "world"
    return result

def empty_except():
    """Empty except block"""
    try:
        result = 10 / 0
    except:
        pass  # Swallowing exception silently

def broad_except():
    """Catching too broad exception"""
    try:
        risky_operation()
    except Exception:  # Too broad
        pass

def very_long_function():
    """Function too long"""
    print("Line 1")
    print("Line 2")
    print("Line 3")
    print("Line 4")
    print("Line 5")
    print("Line 6")
    print("Line 7")
    print("Line 8")
    print("Line 9")
    print("Line 10")
    print("Line 11")
    print("Line 12")
    print("Line 13")
    print("Line 14")
    print("Line 15")
    print("Line 16")
    print("Line 17")
    print("Line 18")
    print("Line 19")
    print("Line 20")
    print("Line 21")
    print("Line 22")
    print("Line 23")
    print("Line 24")
    print("Line 25")
    print("Line 26")
    print("Line 27")
    print("Line 28")
    print("Line 29")
    print("Line 30")
    print("Line 31")
    print("Line 32")
    print("Line 33")
    print("Line 34")
    print("Line 35")
    print("Line 36")
    print("Line 37")
    print("Line 38")
    print("Line 39")
    print("Line 40")

# Bugs

def null_pointer_bug():
    """Null pointer (None) access"""
    my_var = None
    return my_var.upper()  # AttributeError

def division_by_zero():
    """Division by zero"""
    return 10 / 0

def resource_leak():
    """Resource not closed"""
    f = open('test.txt', 'r')
    data = f.read()
    # File never closed

def infinite_loop():
    """Infinite loop"""
    while True:
        print("Loop")
        # No break condition

def unreachable_code():
    """Dead/unreachable code"""
    return 42
    print("This will never execute")
    x = 100

def mutable_default_argument(item, my_list=[]):
    """Mutable default argument bug"""
    my_list.append(item)
    return my_list

def comparison_with_none(value):
    """Wrong None comparison"""
    if value == None:  # Should use 'is'
        return True
    return False

def catching_and_raising():
    """Incorrect exception handling"""
    try:
        risky_operation()
    except Exception as e:
        raise e  # Loses traceback, should use 'raise'

# Bad Practices

def commented_out_code():
    """Commented out code"""
    print("Active code")
    # print("Old code 1")
    # print("Old code 2")
    # print("Old code 3")
    # print("Old code 4")

def todo_comments():
    """TODO comments"""
    # TODO: Fix this later
    # FIXME: This is broken
    # HACK: Temporary solution
    # XXX: Check this
    print("Code")

def print_statements_for_debugging():
    """Debug print statements left in code"""
    x = 42
    print("DEBUG: x =", x)
    print("DEBUG: Processing...")
    return x * 2

# Global variables - bad practice
GLOBAL_VAR = 0
global_counter = 0

def modifying_global():
    """Modifying global variables"""
    global global_counter
    global_counter += 1

# Class with too many methods (God class)
class GodClass:
    def method1(self): pass
    def method2(self): pass
    def method3(self): pass
    def method4(self): pass
    def method5(self): pass
    def method6(self): pass
    def method7(self): pass
    def method8(self): pass
    def method9(self): pass
    def method10(self): pass
    def method11(self): pass
    def method12(self): pass
    def method13(self): pass
    def method14(self): pass
    def method15(self): pass
    def method16(self): pass
    def method17(self): pass
    def method18(self): pass
    def method19(self): pass
    def method20(self): pass

# Empty class
class EmptyClass:
    pass

def risky_operation():
    """Placeholder for risky operation"""
    pass

# Multiple statements on one line
x = 1; y = 2; z = 3

# Wildcard import
from os import *

# Unused import
import json
import sys
import re

# Wrong naming convention
def MyFunction():  # Should be snake_case
    pass

class my_class:  # Should be PascalCase
    pass

CONSTANT_value = 42  # Should be all caps

File: badCode.js
// Security Vulnerabilities

// SQL Injection (using mysql library example)
function sqlInjectionVulnerable(userInput) {
    const mysql = require('mysql');
    const connection = mysql.createConnection({
        host: 'localhost',
        user: 'root',
        password: 'password123', // Hardcoded password
        database: 'mydb'
    });
    
    // SQL Injection - string concatenation
    const query = "SELECT * FROM users WHERE username = '" + userInput + "'";
    connection.query(query, (error, results) => {
        console.log(results);
    });
}

// Hardcoded credentials
const API_KEY = "sk-1234567890abcdef"; // Hardcoded API key
const SECRET_TOKEN = "my_secret_token_12345"; // Hardcoded secret
const PASSWORD = "admin123"; // Hardcoded password

// Using eval - code injection risk
function dangerousEval(userInput) {
    return eval(userInput); // Never use eval with user input
}

// Command injection
function commandInjection(userInput) {
    const exec = require('child_process').exec;
    exec('ls ' + userInput); // Command injection vulnerability
}

// Path traversal
function pathTraversal(fileName) {
    const fs = require('fs');
    const path = './data/' + fileName; // No validation
    return fs.readFileSync(path, 'utf8');
}

// Insecure random
function insecureRandom() {
    return Math.random(); // Not cryptographically secure
}

// XSS vulnerability
function xssVulnerable(userInput) {
    document.getElementById('output').innerHTML = userInput; // XSS risk
}

// Weak cryptography
function weakCrypto(data) {
    const crypto = require('crypto');
    const hash = crypto.createHash('md5'); // MD5 is weak
    hash.update(data);
    return hash.digest('hex');
}

// Code Smells

// Too many parameters
function tooManyParameters(a, b, c, d, e, f, g, h, i) {
    return a + b + c + d + e + f + g + h + i;
}

// High cognitive complexity
function complexFunction(x, y, z) {
    if (x > 0) {
        if (y > 0) {
            if (z > 0) {
                if (x > y) {
                    if (y > z) {
                        if (x > z) {
                            if (x + y > z) {
                                if (x * y > z) {
                                    return true;
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    return false;
}

// Duplicated code block 1
function duplicatedCode1() {
    const name = "John";
    const age = 30;
    const email = "john@example.com";
    console.log("Name: " + name);
    console.log("Age: " + age);
    console.log("Email: " + email);
    console.log("Processing complete");
}

// Duplicated code block 2
function duplicatedCode2() {
    const name = "Jane";
    const age = 25;
    const email = "jane@example.com";
    console.log("Name: " + name);
    console.log("Age: " + age);
    console.log("Email: " + email);
    console.log("Processing complete");
}

// Magic numbers
function calculateTotal(price) {
    const tax = price * 0.18; // Magic number
    const discount = price * 0.15; // Magic number
    const shipping = price * 0.05; // Magic number
    return price + tax - discount + shipping;
}

// Unused variables
function unusedVariables() {
    const unusedVar = 42;
    const anotherUnused = "hello";
    const result = "world";
    return result;
}

// Empty catch block
function emptyCatch() {
    try {
        const result = riskyOperation();
    } catch (error) {
        // Empty catch block
    }
}

// Function too long
function veryLongFunction() {
    console.log("Line 1");
    console.log("Line 2");
    console.log("Line 3");
    console.log("Line 4");
    console.log("Line 5");
    console.log("Line 6");
    console.log("Line 7");
    console.log("Line 8");
    console.log("Line 9");
    console.log("Line 10");
    console.log("Line 11");
    console.log("Line 12");
    console.log("Line 13");
    console.log("Line 14");
    console.log("Line 15");
    console.log("Line 16");
    console.log("Line 17");
    console.log("Line 18");
    console.log("Line 19");
    console.log("Line 20");
    console.log("Line 21");
    console.log("Line 22");
    console.log("Line 23");
    console.log("Line 24");
    console.log("Line 25");
    console.log("Line 26");
    console.log("Line 27");
    console.log("Line 28");
    console.log("Line 29");
    console.log("Line 30");
    console.log("Line 31");
    console.log("Line 32");
    console.log("Line 33");
    console.log("Line 34");
    console.log("Line 35");
    console.log("Line 36");
    console.log("Line 37");
    console.log("Line 38");
    console.log("Line 39");
    console.log("Line 40");
}

// Bugs

// Null pointer bug
function nullPointerBug() {
    let str = null;
    return str.length; // TypeError
}

// Division by zero
function divideByZero(a) {
    return a / 0; // Returns Infinity but still a bug
}

// Unreachable code
function deadCode() {
    return 42;
    console.log("This will never execute"); // Dead code
    const x = 100;
}

// Incorrect comparison
function incorrectComparison(x) {
    if (x = 5) { // Should be === or ==
        return true;
    }
    return false;
}

// Type coercion bugs
function typeCoercionBug() {
    if ("0" == 0) { // Should use ===
        console.log("Equal");
    }
    if (null == undefined) { // Should use ===
        console.log("Also equal");
    }
}

// Missing await
async function missingAwait() {
    const promise = asyncOperation(); // Missing await
    console.log(promise); // Will log Promise object, not result
}

// Infinite loop
function infiniteLoop() {
    while (true) {
        console.log("Loop");
        // No break condition
    }
}

// Variable shadowing
function variableShadowing() {
    const x = 10;
    if (true) {
        const x = 20; // Shadows outer x
        console.log(x);
    }
}

// Not handling promise rejection
function unhandledPromise() {
    fetch('https://api.example.com/data'); // No .catch()
}

// Bad Practices

// Commented out code
function commentedCode() {
    console.log("Active code");
    // console.log("Old code 1");
    // console.log("Old code 2");
    // console.log("Old code 3");
    // console.log("Old code 4");
}

// TODO comments
function todoComments() {
    // TODO: Fix this later
    // FIXME: This is broken
    // HACK: Temporary solution
    console.log("Code");
}

// console.log for debugging
function debugWithConsole() {
    const x = 42;
    console.log("DEBUG: x =", x); // Should use proper logger
    console.log("DEBUG: Processing...");
    return x * 2;
}

// var instead of let/const
function usingVar() {
    var x = 10; // Should use let or const
    var y = 20;
    return x + y;
}

// Multiple var declarations
var a = 1, b = 2, c = 3; // Bad practice

// Callback hell
function callbackHell() {
    asyncOp1(function(result1) {
        asyncOp2(result1, function(result2) {
            asyncOp3(result2, function(result3) {
                asyncOp4(result3, function(result4) {
                    console.log(result4);
                });
            });
        });
    });
}

// Not using strict mode
function noStrictMode() {
    undeclaredVariable = 42; // Should error in strict mode
}

// God object - too many methods
const godObject = {
    method1: function() {},
    method2: function() {},
    method3: function() {},
    method4: function() {},
    method5: function() {},
    method6: function() {},
    method7: function() {},
    method8: function() {},
    method9: function() {},
    method10: function() {},
    method11: function() {},
    method12: function() {},
    method13: function() {},
    method14: function() {},
    method15: function() {},
    method16: function() {},
    method17: function() {},
    method18: function() {},
    method19: function() {},
    method20: function() {}
};

// Using with statement
function usingWith(obj) {
    with (obj) { // 'with' is deprecated
        console.log(property);
    }
}

// Modifying prototype of built-in objects
Array.prototype.myMethod = function() { // Bad practice
    return this.length;
};

// Global variables
var globalVar1 = "global";
globalVar2 = "also global"; // Implicit global

// Function in loop
function functionInLoop() {
    const funcs = [];
    for (var i = 0; i < 10; i++) {
        funcs.push(function() {
            console.log(i); // Will always log 10
        });
    }
    return funcs;
}

// No default case in switch
function noDefaultCase(x) {
    switch(x) {
        case 1:
            return "one";
        case 2:
            return "two";
        // Missing default case
    }
}

// Empty function
function emptyFunction() {
}

// Unused function
function unusedFunction() {
    console.log("Never called");
}

// Placeholder functions
function riskyOperation() {
    throw new Error("Error");
}

function asyncOperation() {
    return Promise.resolve("data");
}

function asyncOp1(callback) {}
function asyncOp2(data, callback) {}
function asyncOp3(data, callback) {}
function asyncOp4(data, callback) {}

// Trailing comma (not always supported)
const objWithTrailingComma = {
    prop1: "value1",
    prop2: "value2",
};

// Multiple statements on one line
var x = 1; var y = 2; var z = 3;
